src/services/prisma-datastore/prismaToSparrow.ts

```diff
@@ -0,0 +1,172 @@
+import Prisma from "@prisma/client";
+import Gateway from "../alpha-models/Gateway";
+import Node from "../alpha-models/Node";
+import { NodeSensorTypeNames } from "../DomainModel";
+import { SignalStrengths } from "../alpha-models/SignalStrengths";
+
+
+export type SensorLatestReadings = (Prisma.Sensor & {
+    latest: Prisma.Reading | null;
+    schema: Prisma.ReadingSchema;
+  })[];
+
+export type LatestReadingSourceReadings = {
+  readingSource: Prisma.ReadingSource & {
+    sensors: SensorLatestReadings
+  }
+};
+
+export type GatewayWithLatestReadings = Prisma.Gateway & LatestReadingSourceReadings;
+export type NodeWithLatestReadings = Prisma.Node & LatestReadingSourceReadings;
+
+
+
+interface LoRaSignalMetrics {
+  rssi: number;
+  snr?: number;
+}
+
+
+function loraSignalMetricsToSignalStrengths(metrics?: LoRaSignalMetrics) : SignalStrengths {
+  if (!metrics) {
+    return "N/A";
+  }
+
+  const bars = loraSignalMetricsToBars(metrics);
+  return barsToSignalStrengths(bars);
+}
+
+function loraSignalMetricsToBars(metrics: LoRaSignalMetrics): number {
+  const bars = rssiToBars(metrics.rssi);
+  const snrOffset = metrics.snr!==undefined ? snrBarsOffset(metrics.snr) : 0;
+  return bars+snrOffset;
+}
+
+function snrBarsOffset(snr: number): number {
+  if (snr<-10) {
+    return -0.5;
+  }
+  else if (snr < 0) {
+    return -0.25;
+  }
+  return 0;
+}
+
+function rssiToBars(rssi: number): number {
+  return (rssi+120)/30;
+}
```

Comment:
the functions are definitely not prisma-specific so should go someplace else.  Maybe presentation or application layer.

Reply:
Maybe a LoRa module in the application layer?

Reply:
moved to SignalStrengths.ts